### DISCUSSION TITLE
`.partial()` leads to undesirable `undefined`

### DISCUSSION THREAD

**Discussion by mrcaidev** [colinhacks/zod#2314] [Q&A] [answered]
Example:

```ts
const schema = z.object({
  a: z.string(),
}).partial();

type T = z.infer<typeof schema>;

// type T = {
//   a?: string | undefined;
// };
```

What I want:

```ts
type T = {
  a?: string;
};
```

...which are the same programmatically, but TypeScript thinks are different. That is, `{ a?: string | undefined }` is not assignable to `Partial<{ a: string }>`, and I have to use `as` to coerce it.

Some screenshots:

This reports error:

![image](https://user-images.githubusercontent.com/78269445/230902529-3cf3c238-e14b-404f-a1da-422e6bbe3a5a.png)

Error detail:

![image](https://user-images.githubusercontent.com/78269445/230902565-4262320c-5ad4-4fe9-8d3b-87f9c1641887.png)

This is correct, but ugly:

![image](https://user-images.githubusercontent.com/78269445/230902624-b34c4135-0457-4927-ad57-e40340c71a48.png)

I have read the docs, issues and discussions but found no elegant solutions. Any ideas? Thanks in advance!

**Comment by yosr-fourati** [comment]
f you want to avoid the undefined value in the inferred type, you can use the built-in NonNullable type to remove it. Here's an example:
```
type T = NonNullable<z.infer<typeof schema>>;

// type T = {
//   a?: string;
// };
```
This type removes the undefined type from T, so the a property is still optional, but it can only have a value of string or null. If you want to disallow the null value as well, you can use the built-in Exclude type like this:
```
type T = Exclude<NonNullable<z.infer<typeof schema>>, null>;

// type T = {
//   a?: string;
// };
```

**Reply by mrcaidev** [reply]
`NonNullable` won't apply to the keys of the schema, but it did remind me that I can define a `DeepNonNullable` type to achieve the result. Thank you!

**Comment by mrcaidev** [comment, marked as answer]
Found a solution:

```ts
import { ZodType, z } from "zod";

type DeepNonNullable<T> = T extends (infer U)[]
  ? DeepNonNullable<U>[]
  : T extends object
  ? { [K in keyof T]: DeepNonNullable<T[K]> }
  : NonNullable<T>;

export type Infer<T extends ZodType> = DeepNonNullable<z.infer<T>>;

const schema = z
  .object({
    a: z.string(),
  })
  .partial();

type T = Infer<typeof schema>;

// type T = {
//   a?: string;
// };
```
